be/src/io/cache/fs_file_cache_storage.cpp

```diff
@@ -326,91 +442,139 @@ Status FSFileCacheStorage::upgrade_cache_dir_if_necessary() const {
     std::string version;
     std::error_code ec;
     int rename_count = 0;
+    int failure_count = 0;
     auto start_time = std::chrono::steady_clock::now();
 
     RETURN_IF_ERROR(read_file_cache_version(&version));
+
     LOG(INFO) << "Checking cache version upgrade. Current version: " << version
               << ", target version: 2.0, need upgrade: "
               << (USE_CACHE_VERSION2 && version != "2.0");
     if (USE_CACHE_VERSION2 && version != "2.0") {
         // move directories format as version 2.0
-        std::filesystem::directory_iterator key_it {_cache_base_path, ec};
-        if (ec) {
-            LOG(WARNING) << "Failed to list directory: " << _cache_base_path
-                         << ", error: " << ec.message();
-            return Status::InternalError("Failed to list dir {}: {}", _cache_base_path,
-                                         ec.message());
-        }
-        for (; key_it != std::filesystem::directory_iterator(); ++key_it) {
-            if (key_it->is_directory()) {
-                std::string cache_key = key_it->path().filename().native();
-                if (cache_key.size() > KEY_PREFIX_LENGTH) {
-                    std::string key_prefix =
-                            Path(_cache_base_path) / cache_key.substr(0, KEY_PREFIX_LENGTH);
-                    bool exists = false;
-                    auto exists_status = fs->exists(key_prefix, &exists);
-                    if (!exists_status.ok()) {
-                        LOG(WARNING) << "Failed to check directory existence: " << key_prefix
-                                     << ", error: " << exists_status.to_string();
-                        return exists_status;
-                    }
-                    if (!exists) {
-                        auto create_status = fs->create_directory(key_prefix);
-                        if (!create_status.ok()) {
-                            LOG(WARNING) << "Failed to create directory: " << key_prefix
-                                         << ", error: " << create_status.to_string();
-                            return create_status;
+        std::vector<std::filesystem::directory_iterator> file_list;
```

Comment:
why do we use iterator as the element type? it may open too many files,
consider keep the file name (string) as the element type